[DataPipe] implementing fork()

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64215
* #63650
* __->__ #63649

Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

cc @VitalyFedyunin @ejguan

NivekT committed:
implementing fork()

[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -586,6 +586,44 @@ def test_concat_datapipe(self):
+        # Test case: one child DataPipe yields all value first
+        dp1, dp2, dp3 = input_dp.fork(3)
+        output1, output2, output3 = [], [], []
+        for n1 in dp1:
```
nit: output1 = list(dp1)

NivekT replied:
Changed.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -586,6 +586,44 @@ def test_concat_datapipe(self):
+            output2.append(n2)
+        for n3 in dp3:
+            output3.append(n3)
+        self.assertEqual(output1, list(range(10)))
```
nit: If im not mistaken the order is `expected, got`

NivekT replied:
I will switch the order.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -586,6 +586,44 @@ def test_concat_datapipe(self):
+        self.assertEqual(output, [(i, i) for i in range(10)])
+
+        # Test case: one child DataPipe yields all value first, but buffer_size = 5 being too small
+        dp1, dp2 = input_dp.fork(2, 5)
```
Make it keyword-only argument please

NivekT replied:
Done.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -1372,19 +1410,22 @@ def test_simple_traverse(self):
-        graph = torch.utils.data.graph.traverse(combined_dp)
-        expected = {combined_dp: {dp0_upd: {dp0: {}}, dp1_upd: {dp1: {}}, dp2: {}}}
-        self.assertEqual(expected, graph)
+    # TODO: This conflicts with new .fork implementation "TypeError: cannot pickle 'generator' object"
```
We need to fix it in this PR

NivekT replied:
I had another look but haven't fix this test yet. I am not 100% sure what the graph is doing, I will look again next week.

NivekT replied:
Fixed.

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
Changes unrelated to fork

NivekT replied:
Removed.

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -55,16 +57,97 @@ def __iter__(self):
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        container = _ForkIterDataPipe(datapipe, num_instances, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
```
This is likely the reason why it is not pickable, consider wrapping it into tuple like `tuple([...bla bla..])`

NivekT replied:
Just tried that and still get the same error. Any suggestion?

```
test/test_datapipe.py:1406 (TestGraph.test_traverse_forked)
Traceback (most recent call last):
  File "/Users/ktse/pytorch/test/test_datapipe.py", line 1414, in test_traverse_forked
    graph = torch.utils.data.graph.traverse(combined_dp)
  File "/Users/ktse/pytorch/torch/utils/data/graph.py", line 43, in traverse
    d[datapipe].update(traverse(item))
  File "/Users/ktse/pytorch/torch/utils/data/graph.py", line 43, in traverse
    d[datapipe].update(traverse(item))
  File "/Users/ktse/pytorch/torch/utils/data/graph.py", line 43, in traverse
    d[datapipe].update(traverse(item))
  File "/Users/ktse/pytorch/torch/utils/data/graph.py", line 40, in traverse
    items = list_connected_datapipes(datapipe)
  File "/Users/ktse/pytorch/torch/utils/data/graph.py", line 34, in list_connected_datapipes
    p.dump(scan_obj)
TypeError: cannot pickle 'generator' object
~``

VitalyFedyunin replied:
In this case advice is to inspect returned objects dicts and see if there are any generators properties

NivekT replied:
We determined this is not the cause of the pickle issue, so I am leaving it untouched.

VitalyFedyunin reviewed (commented):
Please make sure all tests are passing

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -55,16 +57,97 @@ def __iter__(self):
+        as requested by the child DataPipes.
+    """
+    def __init__(self, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        self.main_datapipe = iter(datapipe)
```
no need to call iter here, just use datapipe, you are invoking unnecessary operation here

NivekT replied:
Without calling `iter`, how do I get the next element on line 107 (see below)?

`self.buffer.append(self.main_datapipe.__next__())`

VitalyFedyunin replied:
Sounds reasonable, we will need to take care of reiterating later.

VitalyFedyunin replied:
Actually, this can be the reason of pickle failure, you could try to do lazy initialization of iter(datapipe) inside of get_next

NivekT replied:
Yep, it seems to be the case. I am getting different errors now.

NivekT replied:
I am switching to a lazy initialization of `iter(datapipe)` within the method `get_next` to address the issue with pickle.

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
I have strong feeling that code will work wrongly, as you do popleft all indexes in the buffer are getting moved, but you are not updating child_pointers

VitalyFedyunin replied:
Please add test to cover this logic

NivekT replied:
I compute an index that accounts for the fact that `popleft()` has occurred before accessing the elements in the buffer.

`buffer_index = self.child_pointers[instance_id] - self.slowest_ptr`

I will more some tests to make sure of this.

VitalyFedyunin replied:
Ah, right. You are stacking difference in the slowest_ptr

NivekT committed:
Update on "[DataPipe] implementing fork()"




[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/iter/combining.py:
@VitalyFedyunin The current behavior is that, after the first read through, the subsequent calls to `iter` returns empty.

```
dp1, dp2 = input_dp.fork(2)
print(list(dp1))  # Returns [0, ..., 10]
print(list(dp1))  # Returns []
print(list(dp2))  # Returns [0, ..., 10]
```

If we want subsequent call to read from the main DataPipe again, we will have to decide how the buffer will interact with that. Or perhaps creating a separate object.

VitalyFedyunin replied:
I think the best option for such pipes (with multiple consumers) would be:
1. recreate buffer and update all positions of other consumers to zero
2. if any of the consumers haven't reached the end of buffer (at the moment of reset) - produce warning

NivekT replied:
I implemented this as requested.

NivekT committed:
Update on "[DataPipe] implementing fork()"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] implementing fork()"




[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493945).

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
Wait, does it mean creating iterator from non-exhausted instance will work fine, I think we need to drop this `if` statement so:

```
 a, b = c.fork(2)
 i = iter(a)
 i2 = iter(b)
```
will produce warning

NivekT replied:
The `if` statement means you can only create an iterator if the current instance is exhausted.

NivekT replied:
@VitalyFedyunin Let me know if that is not the desired behavior. There may be a way to allow creating iterator from non-exhausted instance, but I worry that might reset other child iterators.

VitalyFedyunin replied:
If `self.main_datapipe.is_instance_exhausted(self.instance_id)` returns false you are defaulting to do `yield from self.main_datapipe.get_next(self.instance_id)`. 

Meaning if multiple iterators is created without exhausing previous one, they will try to operate on the same buffer position

NivekT replied:
Right, and the ideal behavior should be - if the current iterator is not exhausted, return a new, unrelated one?

In that case, I would probably need to make a copy of the container and reset it.

VitalyFedyunin replied:
No need to create new container. We can define behaviour as: if ANY of the iterators (including self.instance_id) is not exhausted (and at least one of them started consuming items), reset ALL with Warning, else if ALL exhausted, reset ALL.

NivekT replied:
That makes sense. However, if I remove:

```
if self.main_datapipe.is_instance_exhausted(self.instance_id):
```
Then, if someone does:
```
dp1, dp2 = input_dp.fork(2)
i1 = iter(dp1)
...
# Some code here to Iterate half way through i1
...
i2 = iter(dp2)  # This will reset i1, do we want that?
```
Is the behavior above intended? If the user doesn't want that, the user must call `i2 = iter(dp2)` before using the iterator `i1`. Maybe there is an alternative implementation allows `i2` to become an iterator without touching `i1`?

NivekT replied:
Okay sounds good.

NivekT replied:
This seems to create a weird interaction with `mux`. One example is the `_get_pipeline` used in a few test cases

```
 def _get_pipeline(self):
        numbers_dp = NumbersDataset(size=10)
        dp0, dp1 = numbers_dp.fork(num_instances=2)
        dp0_upd = dp0.map(lambda x: x * 10)
        dp1_upd = dp1.filter(lambda x: x % 3 == 1)
        combined_dp = dp0_upd.mux(dp1_upd)
        return combined_dp
```

When the line `__next__` is executed below, it will reset the other pipelines.
```
@functional_datapipe('mux')
class MultiplexerIterDataPipe(IterDataPipe):

    def __init__(self, *datapipes):
        self.datapipes = datapipes

    def __iter__(self):
        iterators = [iter(x) for x in self.datapipes]
        finished = {}
        had_more = True
        while had_more:
            had_more = False
            for i in range(len(iterators)):
                if i not in finished:
                    try:
                        value = iterators[i].__next__()  # Thie line reset other child DataPipes :(
                        had_more = True
                        yield value
                    except StopIteration:
                        finished[i] = 1
```

I tried to mitigate this issue by separating the reset execution and the generator within `fork`
```
    def __iter__(self):
        if not self.main_datapipe.is_every_instance_exhausted():
            warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are now "
                          "resetting the buffer and each child DataPipe will read from the start again.", UserWarning)
        self.main_datapipe.reset()
        # We want to separate the code for reset and yield, so that 'reset' exeutes before __next__ is called
        return self.get_generator_by_id(self.instance_id)

    def get_generator_by_id(self, instance_id: int):
        yield from self.main_datapipe.get_next(self.instance_id)
```

This still doesn't work if `fork` is wrapped by other functions (such as `map` and `filter` in `_get_pipeline`). The reset will not execute until a yield statement is called by the higher functions (e.g. `map`, `filter`. Any suggestion on how to work around this?

NivekT replied:
Added a check to see if the child DataPipe has been read before resetting it.

ejguan commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -55,12 +58,106 @@ def __iter__(self):
+            buffer_size: this restricts how far ahead the leading child DataPipe
+             can read relative to the slowest child DataPipe
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        container = _ForkIterDataPipe(datapipe, num_instances, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
+
+    def __init__(self, *arg):
+        raise Exception("__init__ called instead of __new__")
```
nit:
Do we need to split it into two classes? It seems we can combine them together.
```py
class Fork():
    def __new__(cls, dp, num_instances, buffer_size):
        container = super().__new__(cls)
        container.__init__(dp, num_instances, buffer_size)
        return [ChildDataPipe(container, i) for i in range(num_instances)]
    # Similar to `_ForkIterDataPipe`
    def __init__(self, dp, num_instances, buffer_size):
        ...
```
cc: @VitalyFedyunin

NivekT replied:
Good suggestion, I think this works and I will try this out.

VitalyFedyunin replied:
That will work (and probably more readable)

ejguan replied:
This may not work for the case of multiprocessing as @NivekT just pointed out. Serialization/Deserialization doesn't work as my expectation even with `__getnewargs__` implemented.
I suggest Kevin to keep it as previous. I will do some research on that.

ejguan replied:
Did some research on the reason that it doesn't work well with pickling. Want to summarize my findings:
- When unpickling, __init__ is never called in most of cases
- If `__new__` method is specified, `__getnewargs__` is needed for pickler to pass the arguments to `__new__` method. As shown below, `attribute` in `load` function.
- My proposed method would work with `__new__` method that only returns a single element. But, since we are returning a tuple of elements, at the time of unpickling, the returned tuple of instances (`obj` in below) will be errored out due to missing `__dict__` attribute.
```py
def load(cls, attributes):
    obj = cls.__new__(cls)
    obj.__dict__.update(attributes)
    return obj
```

It turns out the `__new__` and `__init__` have to be splitted into two classes, if we want to enable the multiprocessing with `fork`. Sorry for the noise.

NivekT committed:
Update on "[DataPipe] implementing fork()"


Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493945).

ejguan reviewed (commented):
In general, the logic looks correct. A few comments as below:

Edit: rebasing is needed as a PR to rename several `DataPipe`s including Fork is landed this week.

NivekT committed:
Update on "[DataPipe] implementing fork()"


Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

[ghstack-poisoned]

ejguan commented:
Can you please also add `ForkerIterDataPipe` to https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/__init__.py to expose this API for users?

NivekT committed:
Update on "[DataPipe] implementing fork()"


Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

[ghstack-poisoned]

NivekT commented:
> Can you please also add `ForkerIterDataPipe` to https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/__init__.py to expose this API for users?

Done. Is there anything else that should be changed for approval?

ejguan commented on test/test_datapipe.py:
```diff
@@ -579,6 +579,105 @@ def test_concat_datapipe(self):
+
+        # Test Case: one child DataPipe yields all value first, but buffer_size = 5 being too small
+        dp1, dp2 = input_dp.fork(num_instances=2, buffer_size=5)
+        output1 = []
```
nit: remove output1

ejguan commented on test/test_datapipe.py:
```diff
@@ -579,6 +579,105 @@ def test_concat_datapipe(self):
+            output1.append(n1)
+            output2.append(n2)
+            if i == 4:
+                i1 = iter(dp1)  # Reset both all child DataPipe
```
Is this supposed to raise warning here? If so, can we add a validation here?

ejguan commented on test/test_datapipe.py:
```diff
@@ -579,6 +579,105 @@ def test_concat_datapipe(self):
+        output1, output2 = list(dp1), list(dp2)
+        self.assertEqual(list(range(10)), output1)
+        self.assertEqual(list(range(10)), output2)
+        output1, output2 = list(dp1), list(dp2)
```
ditto

ejguan commented on test/test_datapipe.py:
```diff
@@ -579,6 +579,105 @@ def test_concat_datapipe(self):
+        for i, n3 in enumerate(dp3):
+            output3.append(n3)
+            if i == 4:
+                output1 = list(dp1)  # Resets even though dp3 is only partially read
```
ditto

ejguan commented on test/test_datapipe.py:
```diff
@@ -579,6 +579,105 @@ def test_concat_datapipe(self):
+                break
+        self.assertEqual(list(range(10)), list(dp3))  # dp3 has to read from the start again
+
+        # Test Case: Checking for warning being raised
+        with warnings.catch_warnings(record=True) as wa:
+            dp1, dp2 = input_dp.fork(num_instances=2)
+            output1 = list(dp1)
+            output1 = list(dp1)
+            self.assertEqual(len(wa), 1)
+            self.assertRegex(str(wa[0].message), r"Some child DataPipes are not exhausted")
```
If the warning is checked in previous test cases, we can remove this extra checking.

ejguan reviewed (approved):
Thanks, LGTM. Only few comments toward tests.

NivekT committed:
Update on "[DataPipe] implementing fork()"


Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493945).

NivekT committed:
Update on "[DataPipe] implementing fork()"


Differential Revision: [D30493945](https://our.internmc.facebook.com/intern/diff/D30493945)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493945).
